[llvm-objcopy] Remove empty SHT_GROUP sections

chestnykh opened the pull request:
Currently `llvm-objcopy/llvm-strip` in `--strip-debug` mode doesn't remove such sections. This behavior can lead to incompatibilities with GNU binutils (for examples ld.bfd before https://sourceware.org/PR20520 cannot process the object file contains empty .group section).
The ELF object that contains group section with `.debug_*` sections inside can be obtained by `gcc -g3`.
Fix #97139

MaskRay commented:
> Currently llvm-strip in --strip-debug mode doesn't remove such sections. 

Thanks for noticing this. This applies to other section removal operations like --remove-section, so another test will be useful

chestnykh commented:
> > Currently llvm-strip in --strip-debug mode doesn't remove such sections.
> 
> Thanks for noticing this. This applies to other section removal operations like --remove-section, so another test will be useful

Added some new runs of llvm-strip + llvm-readelf

chestnykh committed:
[llvm-strip] Remove empty SHT_GROUP sections.

Currently llvm-strip in --strip-debug mode doesn't remove such sections.
This behavior can lead to incompatibilities with GNU binutils
(for examples ld.bfd cannot process the object file contains empty .group section).
The ELF object that contains group section
with .debug_* sections inside can be obtained by gcc -g3.
Fix #97139

MaskRay reviewed (commented):
You can reopen this PR...

chestnykh commented on llvm/test/tools/llvm-objcopy/ELF/strip-debug-empty-group.test:
```diff
@@ -0,0 +1,25 @@
+# RUN: yaml2obj %s -o %t
+# RUN: llvm-strip --strip-debug %t -o %t1
+# RUN: llvm-readelf --section-groups %t1 | FileCheck %s
+
```
Moved test to `remove-section-in-group.test`. Added `--remove-section=.debug_macro` test (line 44).

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2206,8 +2206,15 @@ Error Object::removeSections(
   // later.
   std::move(Iter, Sections.end(), std::back_inserter(RemovedSections));
-  // Now finally get rid of them all together.
+  // Now get rid of them all together.
```
Nit: whilst you're modifying this comment, please fix the grammar issue
```suggestion
  // Now get rid of them altogether.
```

chestnykh replied:
done

jh7370 reviewed (commented):
I think it would be worth a test case where you explicitly remove an empty .group section too, to show that the code handles that successfully (i.e. it doesn't try to remove the same .group section twice and blow up).

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/remove-section-in-group.test:
```diff
@@ -35,3 +35,35 @@ Symbols:
+# RUN: llvm-readelf --section-groups %t1 | FileCheck %s --check-prefix=GROUP-REMOVED
+# RUN: llvm-strip --strip-unneeded %t -o %t2
+# RUN: llvm-readelf --section-groups %t2 | FileCheck %s --check-prefix=GROUP-REMOVED
+# RUN: llvm-strip --remove-section=.debug_macro %t -o %t3
```
We don't usually do that in llvm-objcopy testing. The behaviour of `--strip-debug` boils down to `--remove-section=.debug*`, roughly-speaking, so it would just be duplicate testing.

Putting it in other terms, say in 10 years time, somebody comes to look at this test when there are both `--strip-debug` and `--remove-section` cases, they may think there's something special about one or other when in fact the behaviour is generic. This would likely cause confusion and make it harder to understand the code and test purposes.

chestnykh replied:
Thanks for the explanation, got it

chestnykh replied:
@jh7370 done

jh7370 reviewed (approved):
LGTM.

MaskRay commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2206,8 +2206,15 @@ Error Object::removeSections(
+  for (auto Iter = Sections.begin(); Iter != Sections.end(); ++Iter)
+    if (auto GroupSec = dyn_cast<GroupSection>(Iter->get()))
+      if (GroupSec->getMembersCount() == 0)
+        Sections.erase(Iter);
```
This is UB and has a quadratic time complexity if the erased elements are many. https://eel.is/c++draft/vector#modifiers-3

"Effects: Invalidates iterators and references at or after the point of the erase."

You could utilize the return address of erase, but to address the time complexity issue you'd need `erase_if` anyway.

chestnykh replied:
Fixed, thanks!

MaskRay commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -963,6 +963,8 @@ class GroupSection : public SectionBase {
       const DenseMap<SectionBase *, SectionBase *> &FromTo) override;
   void onRemove() override;
 
+  size_t getMembersCount() { return GroupMembers.size(); }
```
`() const`

chestnykh replied:
Fixed

jh7370 reviewed (approved):
LGTM again.
